tests/src/reflection/GenericAttribute/GenericAttributeTests.cs

```diff
@@ -0,0 +1,163 @@
+        AssertAny(b3, a => a is SingleAttribute<int>);
+        AssertAny(b3, a => a is SingleAttribute<bool>);
+        AssertAny(b3, a => (a as MultiAttribute<int>)?.Value == 0);
+        AssertAny(b3, a => (a as MultiAttribute<int>)?.Value == 1);
+        AssertAny(b3, a => (a as MultiAttribute<int>)?.Value == 2);
+        AssertAny(b3, a => (a as MultiAttribute<bool>)?.Value == false);
+        AssertAny(b3, a => (a as MultiAttribute<bool>)?.Value == true);
+
+        var a4 = CustomAttributeExtensions.GetCustomAttributes<SingleAttribute<int>>(programTypeInfo, true);
+        AssertAny(a4, a => a is SingleAttribute<int>);
+
+        var b4 = ((ICustomAttributeProvider)programTypeInfo).GetCustomAttributes(typeof(SingleAttribute<int>), true);
+        AssertAny(b4, a => a is SingleAttribute<int>);
+
+        var a5 = CustomAttributeExtensions.GetCustomAttributes<SingleAttribute<bool>>(programTypeInfo);
+        AssertAny(a5, a => a is SingleAttribute<bool>);
+
+        var b5 = ((ICustomAttributeProvider)programTypeInfo).GetCustomAttributes(typeof(SingleAttribute<bool>), true);
+        AssertAny(b5, a => a is SingleAttribute<bool>);
+
+        var a6 = CustomAttributeExtensions.GetCustomAttributes<MultiAttribute<int>>(programTypeInfo, true);
+        AssertAny(a6, a => (a as MultiAttribute<int>)?.Value == 0);
+        AssertAny(a6, a => (a as MultiAttribute<int>)?.Value == 1);
+        AssertAny(a6, a => (a as MultiAttribute<int>)?.Value == 2);
+
+        var b6 = ((ICustomAttributeProvider)programTypeInfo).GetCustomAttributes(typeof(MultiAttribute<int>), true);
+        AssertAny(b6, a => (a as MultiAttribute<int>)?.Value == 0);
+        AssertAny(b6, a => (a as MultiAttribute<int>)?.Value == 1);
+        AssertAny(b6, a => (a as MultiAttribute<int>)?.Value == 2);
+
+        var a7 = CustomAttributeExtensions.GetCustomAttributes<MultiAttribute<bool>>(programTypeInfo, true);
+        AssertAny(a7, a => (a as MultiAttribute<bool>)?.Value == false);
+        AssertAny(a7, a => (a as MultiAttribute<bool>)?.Value == true);
+
+        var b7 = ((ICustomAttributeProvider)programTypeInfo).GetCustomAttributes(typeof(MultiAttribute<bool>), true);
+        AssertAny(b7, a => (a as MultiAttribute<bool>)?.Value == false);
+        AssertAny(b7, a => (a as MultiAttribute<bool>)?.Value == true);
+
+        var a8 = CustomAttributeExtensions.GetCustomAttributes<MultiAttribute<bool?>>(programTypeInfo, true);
+        AssertAny(a8, a => (a as MultiAttribute<bool?>)?.Value == null);
+
+        var b8 = ((ICustomAttributeProvider)programTypeInfo).GetCustomAttributes(typeof(MultiAttribute<bool?>), true);
+        AssertAny(b8, a => (a as MultiAttribute<bool?>)?.Value == null);
+
+        Assert(CustomAttributeExtensions.GetCustomAttributes(programTypeInfo, typeof(MultiAttribute<>), false) == null);
+        Assert(CustomAttributeExtensions.GetCustomAttributes(programTypeInfo, typeof(MultiAttribute<>), true) == null);
```

Comment:
If you pass in an open attribute type to this api (`typeof(MultiAttribute<)`), I think it should either:

- Throw an ArgumentException saying explicitly that the attribute type cannot be an open type (and for consistency, make Attribute.GetCustomAttributes work the same way) I think this is the preferable option as I can't think of a good use case for passing in an open type here.

or 

- Return a zero-length array of type `Attribute[]`.

We shouldn't be returning `null` - there is no precedent for this api returning `null`.

Reply:
@AtsushiKan I implemented this on Attribute.GetCustomAttributes and with CustomAttributeExtensions api it throws NotSupportedException as expected
but it not work for ICustomAttributeProvider api is this api need also to throw?